sdk/iot/Azure.Iot.Hub.Service/tests/QueryTests.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading.Tasks;
+using Azure.Iot.Hub.Service.Models;
+using FluentAssertions;
+using NUnit.Framework;
+
+namespace Azure.Iot.Hub.Service.Tests
+{
+    /// <summary>
+    /// Test the query API. This API is a query on Twins only.
+    /// </summary>
+    /// <remarks>
+    /// All API calls are wrapped in a try catch block so we can clean up resources regardless of the test outcome.
+    /// </remarks>
+    public class QueryTests : E2eTestBase
+    {
+        public QueryTests(bool isAsync)
+            : base(isAsync)
+        {
+        }
+
+        /// <summary>
+        /// Test querying all device twins in the IoTHub.
+        /// </summary>
+        [Test]
+        public async Task QueryClient_GetAllDevicesTwinsWithTag()
+        {
+            string testDeviceId = $"QueryDevice{GetRandom()}";
+
+            DeviceIdentity device = null;
+            IotHubServiceClient client = GetClient();
+
+            try
+            {
+                // Create a device.
+                device = (await client.Devices
+                    .CreateOrUpdateIdentityAsync(
+                        new DeviceIdentity
+                        {
+                            DeviceId = testDeviceId
+                        })
+                    .ConfigureAwait(false))
+                    .Value;
+
+                // Adding a delay to account for query cache sync.
+                // TODO: Figure out a better way to ensure the test always succeeds.
+                await Task.Delay(TimeSpan.FromSeconds(5)).ConfigureAwait(false);
```

Comment:
In the track 1 repo, we typically re-query until the results match our expectations or it hits a timeout. Thread sleeps like this one led to flakey tests against lower tier hubs when combined with lots of other tests

Reply:
Makes sense. I thought about retries too but was not sure what was the best option. I agree retry is better than a single delay. I will check this in and make a change to address this and use retries.